<a href="https://colab.research.google.com/github/Khalil-Salama/Prediction-of-Product-Sales/blob/main/Prediction_of_Product_Sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Title:(Prediction of Product Sales)

- Auther: Khalil Salama

## Project Overview

## Load and Inspect Data

*Mount Google drive*

In [132]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


*Import required packges*

In [133]:
# Import required packages
import pandas as pd
sales_pred=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/AXSOSACADEMY/01-Fundamentals/Week03/Data/sales_predictions_2023.csv')
sales_pred

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

*Display summery information*

In [134]:
# The first five rows
sales_pred.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [135]:
# General information
sales_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


    After reviewing the information there is no obvoius changes, moving forword to checking and cleaning data


## Cleaning Data

*Number of rows*
- 8523

*Number of columns*
- 12

 *Data types for each variable*

In [136]:
sales_pred.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

**Duplicates**

*Finding duplicates*

In [156]:
# finding duplicates
sales_pred_duplicates=sales_pred.duplicated()
sales_pred_duplicates.sum()

np.int64(0)

    No duplicates where found, moving to check unique values

- Checking unique values


In [139]:
sales_pred.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [140]:
# checking the uniquie percentge for each column
sales_pred.nunique()/len(sales_pred)

Item_Identifier              0.182917
Item_Weight                  0.048692
Item_Fat_Content             0.000587
Item_Visibility              0.924557
Item_Type                    0.001877
Item_MRP                     0.696703
Outlet_Identifier            0.001173
Outlet_Establishment_Year    0.001056
Outlet_Size                  0.000352
Outlet_Location_Type         0.000352
Outlet_Type                  0.000469
Item_Outlet_Sales            0.409832
dtype: float64

In [141]:
# checking the unique value in Item Visibility
sales_pred['Item_Visibility'].value_counts()

Item_Visibility
0.000000    526
0.076975      3
0.061271      2
0.046825      2
0.112399      2
           ... 
0.054142      1
0.018920      1
0.044974      1
0.103611      1
0.083060      1
Name: count, Length: 7880, dtype: int64

*Finding missing Values*

In [142]:
# Misssing values
null_sums=sales_pred.isna().sum()
null_columns= null_sums[null_sums>0]
null_columns

Item_Weight    1463
Outlet_Size    2410
dtype: int64

 *Adressing missing floats*


In [143]:
# Knowing the type of columns
sales_pred[null_columns.index].dtypes

Item_Weight    float64
Outlet_Size     object
dtype: object

In [144]:
# adressing the nulls in Float
null_item_weight= sales_pred['Item_Weight'].isna()
null_item_weight.sum()

np.int64(1463)

In [145]:
# group by Item weight to check outlet sales change
sales_pred.groupby(['Item_Weight'])['Item_Outlet_Sales'].mean()

Item_Weight
4.555     1761.706800
4.590     1267.683200
4.610     1993.595429
4.615     4078.025000
4.635     2184.489800
             ...     
21.000    1194.778100
21.100    1851.472306
21.200    2763.735800
21.250    2199.192883
21.350    3391.490086
Name: Item_Outlet_Sales, Length: 415, dtype: float64

    After looking at with Outlet Sales
    and since Item weight dosen't effect the value of outlet sales item.
    Decided to replace it with the mean figure -1


In [146]:
# replacing the null in item weight with -1
sales_pred['Item_Weight']=sales_pred['Item_Weight'].fillna(-1)
# confirming no nulls any more
sales_pred['Item_Weight'].isna().sum()

np.int64(0)

*Adressing missing strings*

In [147]:
# Checking the type of columns
sales_pred[null_columns.index].dtypes

Item_Weight    float64
Outlet_Size     object
dtype: object

In [148]:
# grouping outlet sales by oulet size
sales_pred.groupby(['Outlet_Size'])['Item_Outlet_Sales'].mean()

Outlet_Size
High      2298.995256
Medium    2681.603542
Small     1912.149161
Name: Item_Outlet_Sales, dtype: float64

In [149]:
sales_pred['Outlet_Size'].mode()

0    Medium
Name: Outlet_Size, dtype: object

    After looking at the Outlet Size values decided to replace the nulls with Meduim (the mode of this column)

In [150]:
# replacing null in Outlet size with Meduim
sales_pred['Outlet_Size']=sales_pred['Outlet_Size'].fillna('Medium')
# confirming no nulls any more
sales_pred['Outlet_Size'].isna().sum()

np.int64(0)

*Finding and fixing any inconsistent categories of data.*


In [151]:
# filtering the string columns
string_columns=sales_pred.select_dtypes(include='object').columns
string_columns

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [152]:
# making a for loop fpor the string columns
for col in string_columns:
  print(f"{col}:")
  print(sales_pred[col].value_counts(dropna=False))
  print("\n")

Item_Identifier:
Item_Identifier
FDW13    10
FDG33    10
FDX31     9
FDT07     9
NCY18     9
         ..
FDO33     1
FDK57     1
FDT35     1
FDN52     1
FDE52     1
Name: count, Length: 1559, dtype: int64


Item_Fat_Content:
Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: count, dtype: int64


Item_Type:
Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: count, dtype: int64


Outlet_Identifier:
Outlet_Identifier
OUT027    935
OUT013    932
OUT035    930
OUT049    930
OUT046    930
OUT045    929


    Found inconsistancy in column Item_Fat_Contant

In [153]:
# fixing the inconsistancy in Item fat contant
sales_pred['Item_Fat_Content']=sales_pred['Item_Fat_Content'].replace({'LF':'Low Fat','reg':'Regular','low fat':'Low Fat'})

*Obtaining the summary statistics (min, max, mean) for each numerical column.*

In [154]:
# filtering the numeric columns
num_cols=sales_pred.select_dtypes(include='number').columns
num_cols

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year', 'Item_Outlet_Sales'],
      dtype='object')

In [155]:
# Making a for loop for all numeric columns to get thier min, max, and mean
for col in num_cols:
  print(f"{col}:")
  print(f"min: {sales_pred[col].min()}")
  print(f"max: {sales_pred[col].max()}")
  print(f"mean: {sales_pred[col].mean()}")

Item_Weight:
min: -1.0
max: 21.35
mean: 10.478936407368296
Item_Visibility:
min: 0.0
max: 0.328390948
mean: 0.06613202877895108
Item_MRP:
min: 31.29
max: 266.8884
mean: 140.9927819781767
Outlet_Establishment_Year:
min: 1985
max: 2009
mean: 1997.8318667135984
Item_Outlet_Sales:
min: 33.29
max: 13086.9648
mean: 2181.288913575032


## Exploratory Data Analysis

## Feature Inspection